docs/source/gbd2019_models/hemoglobin_anemia_and_iron_deficiency/hemoglobin/index.rst

```diff
@@ -174,69 +174,16 @@ Data Description Tables
 
 Below is R code written to randomly sample hemoglobin concentration values from the hemoglobin distribution parameters and constants defined in the tables above. Additionally, the code block contains functions that will evaluate the proportion of the distribution below a given threshold. This code was adapted from the GBD stash code found `here <https://stash.ihme.washington.edu/projects/MNCH/repos/anemia/browse/model/envelope>`__, specifically the *DistList_mv2p.R* and *fit_ensemblemv2p_parallel.R* files.
 
-.. note::
-
-  This code has been translated to python in a notebook hosted `here <https://github.com/ihmeuw/vivarium_gates_lsff/blob/main/tests/lsff_iron_exposure.ipynb>`__.
+**Code to sample from the hemoglobin exposure distribution in python has been developed and hosted in a notebook hosted** `here <https://github.com/ihmeuw/vivarium_gates_lsff/blob/main/tests/lsff_iron_exposure.ipynb>`__.
 
 .. todo::
 
-	Integrate python code from above notebook into this document.
-
-.. code-block:: R
-
-  # define constants
-  XMAX = 220
-  EULERS_CONSTANT = 0.57721566490153286060651209008240243104215933593992
-  w = c(0.4,0.6)
-
-  # import standard R functions for the gamma distributions (pgamma and rgamma)
-  pacman::p_load(data.table,actuar)
-
-  # function to calculate gamma distribution parameters from mean and variance
-  gamma_mv2p = function(mn, vr){
-    list(shape = mn^2/vr,rate = mn/vr)}
-
-  # function to calculate mirror gumbel distribution parameters from mean and variance 
-  mgumbel_mv2p = function(mn, vr){
-    list(
-      alpha = XMAX - mn - EULERS_CONSTANT*sqrt(vr)*sqrt(6)/pi,
-      scale = sqrt(vr)*sqrt(6)/pi)}
-  # function to randomly sample n times from mirror gumbel distribution
-  rmgumbel = function(n, alpha, scale){
-    mn = alpha + scale*EULERS_CONSTANT
-    rgumbel(n, alpha+XMAX-(2*mn), scale)}
-
-  # function to calculate area under curve below threshold q for mirror gumbel distribution
-  pmgumbel = function(q, alpha, scale, lower.tail){ 
-    #NOTE: with mirroring, take the other tail
-    pgumbel(XMAX-q, alpha, scale, lower.tail=ifelse(lower.tail,FALSE,TRUE))}
-
-  # function to calculate area under curve of hemoglobin ensemble distribution using the functions defined above
-    # q = hemoglobin threshold
-    # mn = mean hemoglobin concentration
-    # vr = hemoglobin variance (standard deviation squared)
-    # w = list of ensemble distribution weights c(gamma_weight, mirror_gumbel_weight)
-  ens_mv2prev <- function(q, mn, vr, w){
-    x = q
-
-    ##parameters
-    params_gamma = gamma_mv2p(mn, vr)
-    params_mgumbel = mgumbel_mv2p(mn, vr)
-
-    ##weighting
-    prev = sum(
-      w[1] * pgamma(x, params_gamma$shape, params_gamma$rate), 
-      w[2] * pmgumbel(x, params_mgumbel$alpha, params_mgumbel$scale, lower.tail=T))
-    prev}
+  Determine if this code follows the protocol to appropriately sample from an ensemble distribution, as identified by Nathaniel.
```

Comment:
Yes, Kjell implemented the corrected sampling algorithm I suggested, though the notebook also has the original sampling method, so it probably won't be clear to the reader what they're looking for. I think it would be good to copy the relevant functions into the documentation (and perhaps give them better names):
```Python
import numpy as np
import pandas as pd
import scipy

def _gamma_ppf(propensity, mean, sd):
"""Returns the quantile of a gamma distribution with the specified mean and
standard deviation corresponding to the specified quantile rank (propensity).
"""
    shape = (mean / sd)**2
    scale = sd**2 / mean
    return scipy.stats.gamma(a=shape, scale=scale).ppf(propensity)

def _mirrored_gumbel_ppf(propensity, mean, sd):
"""Returns the quantile of a mirrored Gumbel distribution with the specified mean
and standard deviation corresponding to the specified quantile rank (propensity).
"""
    x_max = 220 # data_values.HEMOGLOBIN_DISTRIBUTION.EXPOSURE_MAX
    _alpha = x_max - mean - (sd * np.euler_gamma * np.sqrt(6) / np.pi)
    scale = sd * np.sqrt(6) / np.pi
    tmp = _alpha + (scale*np.euler_gamma)
    alpha = _alpha + x_max - (2*tmp)
    return scipy.stats.gumbel_r(alpha, scale=scale).ppf(propensity)

# Called `viv_calc_iron_nbs` in Kjell's notebook
def sample_from_hemoglobin_distribution(prop_dist, propensity, exposure_parameters):
"""
Returns a sample from an ensemble distribution with the specified mean and
standard deviation (stored in `exposure_parameters`) that is 40% Gamma and
60% mirrored Gumbel. The sampled value is a function of the two propensities
`prop_dist` (used to choose whether to sample from the Gamma distribution or
the mirrored Gumbel distribution), and `propensity` (used as the quantile rank
for the selected distribution).
"""
    #propensity = clip(propensity)
    exposure_data = exposure_parameters
    mean = exposure_data['mean']
    sd = exposure_data['sd']

    gamma = prop_dist < 0.4
    gumbel = ~gamma
    ret_val = pd.Series(index=prop_dist.index, name='value')
    ret_val.loc[gamma] = _gamma_ppf(propensity.loc[gamma], mean, sd)
    ret_val.loc[gumbel] = _mirrored_gumbel_ppf(propensity.loc[gumbel], mean, sd)
    return ret_val
```
Note that the corrected version of ensemble sampling has already been incorporated into Vivarium's `risk_distribution` module in the [.ppf method of the EnsembleDistribution class](https://github.com/ihmeuw/risk_distributions/blob/main/src/risk_distributions/risk_distributions.py#L495) (which is misleadingly named because this sampling method is not the same thing as the quantile function...).